### ISSUE TITLE
Design

### ISSUE THREAD

**Issue by Avsecz** [kipoi/dataloaders#1]
Which datalaoders to use:
- DNA sequence based
- bigiwig-based (useful for DNA + accessibility)
- Efficient bed-file reader

## API
- use `torch.data.Dataset` API

## Nice to have
- transforms
  - reverse-complementation

## Integration with Kipoi
- allo

**Comment by lzamparo** [comment]
Two suggestions: 
- from our previous thread (#330 in kipoi/kipoi)  re: DeepBind dataloaders, I would +1  pyfaidx for dealing with fasta indexing.  It has fewer dependencies than GenomeLake (sp?).
-  the actual encoding from concise was good, but maybe could be factored out of the package?  Installing the whole concise package just to encode DNA was a bit overkill

**Comment by Avsecz** [comment]
Agree. We should also add proper handling of variable widths etc. 

I will indeed factor out the dataloader from concise for the beginning to make things simple. It will not be as fast as genomelake but it's a good start.

**Comment by Avsecz** [comment]
@lzamparo what was your issue once again with genomelake - pysam?

**Comment by Avsecz** [comment]
Migrating from pysam to pyfaidx: https://github.com/kundajelab/genomelake/pull/19

**Comment by krrome** [comment]
thanks for the discussion - I am currently implementing a first draft of the dataloaders repo and I think my suggestion agrees with many of your points. I have opted to:

- write my own `FastaExtractor` class and called it `FastaReader` to avoid confusion. This is necessary as some models (e.g: lsgkm-SVM) need string inputs, others one-hot encoded input, etc. It is therefore a wrapper around pyfaidx so that it operates with `pybedtools.Interval` objects.
- the string sequence output can be converted by `transformer` methods, one of which is the `onehot_trafo`, another will be `char_list_trafo`: 
   - `onehot_trafo` is at the moment a three-liner which can at any point be replaced with the `onehot` method in genomelake. I didn't do that to avoid genomelake dependency only because of this function
   - `char_list_trafo` converts the sequence string returned by `FastaReader` into list of characters, as required by some models.
- For one-hot encoded sequences I define a Reshaper class that can add "dummy" dimensions and swap the order of the sequence and alphabet axes as required by the model.

using these definitions, I get a neat and compact defintion of dataloader classes capable to cater for one-hot encoded models with all sorts of different input shapes and axis-order, string-sequece models, etc. see: https://github.com/kipoi/dataloaders/blob/first_draft/kipoi_dataloaders/sequence_based.py. please feel free to take a look at the branch https://github.com/kipoi/dataloaders/tree/first_draft and give feedback :)

**Comment by krrome** [comment]
@lzamparo btw. I am more than happy to use a fasta reading package that has the functionality you requested. pyfaidx does not have the functionality that you have requested - e.g.: requesting the sequence from genomic interval chr15:98503515-98503517 given the fasta example file in your [gist](https://gist.github.com/lzamparo/cdeb95fd0dd43a32eeefb8aa6eeb0cb3) in https://github.com/kipoi/kipoi/issues/330. pyfaidx inteprets the entire string after the ">" as a sequence/chromosome name. If you can point me to the fasta reader that you have in mind which is capable to handle this the way you specified and if the that fasta reader complies with the fasta standards then I am more than happy to replace pyfaidx with that one.

**Comment by lzamparo** [comment]
@Avsecz Here's a sample .fasta file in a [gist](https://gist.github.com/lzamparo/5635881af4c01f056de21bbfd2f2aea5), along with the test interval and expected output

**Comment by lzamparo** [comment]
@krrome I've not used `get_seq` before, but I used the dict interface of `Pyfaidx.Fasta` to do slicing on Fasta records from a file organized like the one in my original gist you reference.  

There is one snag, which is you need to convert to relative indexing of the bases from the absolute indexing of the interval, but that's pretty fast.  Have a look at the `TruncSeqDataset` dataloader [I made](https://raw.githubusercontent.com/lzamparo/bindspace_revisions/master/deepbind/src/dataloader.py), you can see how I'm taking subsequences in `__getitem__`

**Comment by krrome** [comment]
@lzamparo thanks for the link. I could add that way of querying in case the default get_seq throws a `FetchError`

**Comment by lzamparo** [comment]
> .. but the problem I can see there is that it will only work if the bed query is matched by a fasta file entry...

True, I ensured there would be correspondance by construction.  Though I think it can handle the .other common use case, where you have one fasta file per chromosome, and the intervals can be taken as a subset without any conversion to relative sequences.  Up to you to decide how to handle the case where an interval has no match in the fasta provided.

**Comment by Avsecz** [comment]
Great! Thanks @lzamparo . I'll have a closer look at your implementation tomorrow and add the provided files to unit-tests. 

## Update
- I'm almost done writing the `transforms` module that contains all the useful pre-processing functions like one-hot encoding, tokenizing or sequence trimming. I re-factored one-hot-encoding functions from concise and also imported the fast one-hot-encoding from genomelake (for the DNA case)
  - @krrome @lzamparo do you frequently use any other functions in addition to those in https://github.com/kipoi/kipoiseq/blob/master/kipoiseq/transforms/functional.py?
    - I was thinking of also adding a k-mer or gapped k-mer counter and reverse complementation function

**Comment by lzamparo** [comment]
@Avsecz +1 for both transforms you suggest below, I can think of some handy use-cases for both.  Also useful might be a random substring transform?  I can imagine a scenario when you are working with longer sequences, but the property you're trying to predict is valid over the length of the sequence, and you want to use this for data augmentation.
